cmd/flarectl/firewall.go

```diff
@@ -0,0 +1,380 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/codegangsta/cli"
+)
+
+func firewallAccessRules(c *cli.Context) {
+	if err := checkEnv(); err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	organizationID, zoneID, err := getScope(c)
+	if err != nil {
+		return
+	}
+
+	// Create an empty access rule for searching for rules
+	rule := cloudflare.AccessRule{
+		Configuration: getConfiguration(c),
+	}
+	if c.String("scope-type") != "" {
+		rule.Scope.Type = c.String("scope-type")
+	}
+	if c.String("notes") != "" {
+		rule.Notes = c.String("notes")
+	}
+	if c.String("mode") != "" {
+		rule.Mode = c.String("mode")
+	}
+
+	var rules []cloudflare.AccessRule
+	if organizationID != "" {
+		response, err := api.ListOrganizationAccessRules(organizationID, rule, 1)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		totalPages := response.ResultInfo.TotalPages
+		rules = make([]cloudflare.AccessRule, 0, response.ResultInfo.Total)
+		rules = append(rules, response.Result...)
+		if totalPages > 1 {
+			for page := 2; page < totalPages; page++ {
+				response, err := api.ListOrganizationAccessRules(organizationID, rule, page)
+				if err != nil {
+					fmt.Println(err)
+					return
+				}
+				rules = append(rules, response.Result...)
+			}
+		}
+	} else if zoneID != "" {
+		response, err := api.ListZoneAccessRules(zoneID, rule, 1)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		totalPages := response.ResultInfo.TotalPages
+		rules = make([]cloudflare.AccessRule, 0, response.ResultInfo.Total)
+		rules = append(rules, response.Result...)
+		if totalPages > 1 {
+			for page := 2; page < totalPages; page++ {
+				response, err := api.ListZoneAccessRules(zoneID, rule, page)
+				if err != nil {
+					fmt.Println(err)
+					return
+				}
+				rules = append(rules, response.Result...)
+			}
+		}
+	} else {
+		response, err := api.ListUserAccessRules(rule, 1)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		totalPages := response.ResultInfo.TotalPages
+		rules = make([]cloudflare.AccessRule, 0, response.ResultInfo.Total)
+		rules = append(rules, response.Result...)
+		if totalPages > 1 {
+			for page := 2; page < totalPages; page++ {
+				response, err := api.ListUserAccessRules(rule, page)
+				if err != nil {
+					fmt.Println(err)
+					return
+				}
+				rules = append(rules, response.Result...)
+			}
+		}
+	}
+	var output []table
```

Comment:
`output := make([]table, 0, len(rules))` may avoid several expensive heap allocations and copies (although I have no idea if that's likely to matter or how long the output generally is).

Reply:
Will fix.